api/team.js

```diff
@@ -0,0 +1,144 @@
+import { renderStatsCard } from "../src/cards/stats-card.js";
+import { blacklist } from "../src/common/blacklist.js";
+import {
+  clampValue,
+  CONSTANTS,
+  parseArray,
+  parseBoolean,
+  renderError,
+} from "../src/common/utils.js";
+import { fetchStats } from "../src/fetchers/stats-fetcher.js";
+import { isLocaleAvailable } from "../src/translations.js";
+import { rank_to_level } from "../src/constant.js";
+
+export default async (req, res) => {
+  const {
+    usernames,
+    hide,
+    hide_title,
+    hide_border,
+    card_width,
+    hide_rank,
+    show_icons,
+    include_all_commits,
+    line_height,
+    title_color,
+    ring_color,
+    icon_color,
+    text_color,
+    text_bold,
+    bg_color,
+    theme,
+    cache_seconds,
+    exclude_repo,
+    custom_title,
+    locale,
+    disable_animations,
+    border_radius,
+    number_format,
+    border_color,
+    rank_icon,
+    show,
+    team_name,
+  } = req.query;
+  res.setHeader("Content-Type", "image/svg+xml");
+
+  const username_list = usernames.split(",");
+
+  for (const username of username_list) {
+    if (blacklist.includes(username)) {
+      return res.send(renderError("Something went wrong"));
+    }
+  }
+  if (locale && !isLocaleAvailable(locale)) {
+    return res.send(renderError("Something went wrong", "Language not found"));
+  }
+
+  const team_stats = {
+    name: "",
+    totalPRs: 0,
+    totalPRsMerged: 0,
+    mergedPRsPercentage: 0,
+    totalReviews: 0,
+    totalCommits: 0,
+    totalIssues: 0,
+    totalStars: 0,
+    totalDiscussionsStarted: 0,
+    totalDiscussionsAnswered: 0,
+    contributedTo: 0,
+    rank: { level: "C", percentile: 0 },
+  };
+
+  try {
+    for (let username of username_list) {
+      const stats = await fetchStats(
+        username,
+        parseBoolean(include_all_commits),
+        parseArray(exclude_repo),
+      );
+      team_stats.totalPRs += stats.totalPRs;
+      team_stats.totalPRsMerged += stats.totalPRsMerged;
+      team_stats.mergedPRsPercentage += stats.mergedPRsPercentage;
+      team_stats.totalReviews += stats.totalReviews;
+      team_stats.totalCommits += stats.totalCommits;
+      team_stats.totalIssues += stats.totalIssues;
+      team_stats.totalStars += stats.totalStars;
+      team_stats.totalDiscussionsStarted += stats.totalDiscussionsStarted;
+      team_stats.totalDiscussionsAnswered += stats.totalDiscussionsAnswered;
+      team_stats.contributedTo += stats.contributedTo;
+      team_stats.rank.percentile += stats.rank.percentile;
+    }
```

Comment:
@pursonchen I think that the following lines of code should be totally reworked if we are going to merge this feature someday. I expect to have only one GraphQL request which will fetch required data about team members. Calling `fetchStats()` function inside loop is an error since it decrease performace and cause hitting upstream GitHub API rate limits. Algorithm complexity should be improved from $O(n)$ to $O(1)$ as previously noticed @rickstaa.

I also do not see much sense in showing sum stats of n people and their average rank. Is showing rank useful in case of team card? I do not think so. I think it is a place for discussion.

Remember that you can implement anything and deploy it on your own Vercel instance.

Reply:
@qwerty541, excellent point! It seems that with [GraphQL v4](https://docs.github.com/en/graphql), data for multiple users can now be fetched without using more GraphQL points:

```graphql
{
  user1: user(login: "rickstaa") {
    ...UserFragment
  }
  user2: user(login: "qwerty541") {
    ...UserFragment
  }
  user3: user(login: "pursonchen") {
    ...UserFragment
  }
  rateLimit {
    limit
    cost
    remaining
    resetAt
  }
}

fragment UserFragment on User {
  login
  name
  contributionsCollection {
    totalCommitContributions
    totalPullRequestReviewContributions
  }
  repositoriesContributedTo(
    first: 1
    contributionTypes: [COMMIT, ISSUE, PULL_REQUEST, REPOSITORY]
  ) {
    totalCount
  }
  pullRequests(first: 1) {
    totalCount
  }
  mergedPullRequests: pullRequests(states: MERGED) {
    totalCount
  }
  openIssues: issues(states: OPEN) {
    totalCount
  }
  closedIssues: issues(states: CLOSED) {
    totalCount
  }
  followers {
    totalCount
  }
  repositoryDiscussions {
    totalCount
  }
  repositoryDiscussionComments(onlyAnswers: true) {
    totalCount
  }
}
```

I was surprised to learn about that feature and that it only uses 1 GraphQL point. Thanks for bringing it to my attention. As a result, making the addition of this card possible without having to wait for the GitHub action to be released. 

However, creating a separate team data fetcher is essential so that the codebase doesn't get cluttered. Furthermore, we have to discuss which stats would be interesting to show for teams. I think the following stats are not helpful when looking at teams:

- Contributed to (last year).
- Total discussions started.
- Total discussion answered.
- PR reviews.